keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,535 @@
+-->
+
+- Existing Pod Lifecycle tests must pass fine even after increasing the relisting frequency.
+
+
+### Graduation Criteria
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Existing `node e2e` tests around pod lifecycle must pass
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A.
+
+Since this feature alters only the way kubelet determines the container statuses, this section is irrelevant to this feature.
+
+## Production Readiness Review Questionnaire
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: EventedPLEG
+  - Components depending on the feature gate: kubelet
+
+###### Does enabling the feature change any default behavior?
+
+This feature does not introduce any user facing changes. Although users should notice increased performance of the kubelet which should result in reduced overhead of kubelet and the CRI runtime after enabling this feature.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, kubelet needs to be restarted to disable this feature.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+If reenabled, kubelet will again start updating container statuses using CRI events instead of relisting. Everytime this feature is enabled or disabled, the kubelet will need to be restarted. Hence, the kubelet will start from a clean state.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, unit tests for the feature when enabled and disabled will be implemented in both kubelet
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+This feature relies on the CRI runtime events to determine the container statuses. If the CRI runtime is not upgraded to the version which emits those CRI events before enabling this feature, the kubelet will not be able to determine the container statuses immediately. However, we aren't getting rid of the exiting relisting altogether. So the kubelet should eventually reconcile the container statuses using relisting abeit rather more infrequently due to [increased relisting period](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162) that comes with this feature.
```

Comment:
for CRI stats KEP, we ended up falling back to the old behavior if the CRI didn't support the new stats. Perhaps we should check if the CRI supports the events, and if not, return to the higher frequency (this will help pass alpha e2e tests before CRI supports it)

Reply:
Thanks, addressed it in https://github.com/kubernetes/enhancements/pull/3387/files#diff-4a48c7915d4769b7522c1c9711a927f4ebacfcefe8b2b60c9edf6ffd43e35498R456